Mathlib/Analysis/Fourier/AddCircleMulti.lean

```diff
@@ -0,0 +1,263 @@
+    refine adjoin_le ?_
+    rintro _ ⟨n, rfl⟩
+    refine subset_adjoin ⟨-n, ?_⟩
+    ext1 x
+    simp only [mFourier_neg, starRingEnd_apply, ContinuousMap.star_apply]
+
+/-- The star subalgebra of `C(UnitAddTorus d, ℂ)` generated by `fourier n` for `n ∈ ℤᵈ` is in fact
+the linear span of these functions. -/
+theorem mFourierSubalgebra_coe :
+    (mFourierSubalgebra d).toSubalgebra.toSubmodule = span ℂ (range mFourier) := by
+  apply adjoin_eq_span_of_subset
+  refine .trans (fun x ↦ Submonoid.closure_induction (fun _ ↦ id) ⟨0, ?_⟩ ?_) subset_span
+  · ext z
+    simp only [mFourier, Pi.zero_apply, fourier_zero, Finset.prod_const, one_pow,
+      ContinuousMap.coe_mk, ContinuousMap.one_apply]
+  · rintro _ _ _ _ ⟨m, rfl⟩ ⟨n, rfl⟩
+    refine ⟨m + n, ?_⟩
+    ext z
+    simp only [mFourier, Pi.add_apply, fourier_apply, fourier_add', Finset.prod_mul_distrib,
+      ContinuousMap.coe_mk, ContinuousMap.mul_apply]
+
+/-- The subalgebra of `C(UnitAddTorus d, ℂ)` generated by `mFourier n` for `n : d → ℤ` separates
+points. -/
+theorem mFourierSubalgebra_separatesPoints [DecidableEq d] :
+    (mFourierSubalgebra d).SeparatesPoints := by
+  intro x y hxy
+  rw [Ne, funext_iff, not_forall] at hxy
+  obtain ⟨i, hi⟩ := hxy
+  refine ⟨_, ⟨mFourier (Pi.single i 1), subset_adjoin ⟨Pi.single i 1, rfl⟩, rfl⟩, ?_⟩
+  dsimp only
+  rw [mFourier_single, mFourier_single, fourier_one, fourier_one, Ne, Subtype.coe_inj]
+  contrapose! hi
+  exact AddCircle.injective_toCircle one_ne_zero hi
+
+/-- The subalgebra of `C(UnitAddTorus d, ℂ)` generated by `mFourier n` for `n : d → ℤ` is dense. -/
+theorem mFourierSubalgebra_closure_eq_top [DecidableEq d] :
+    (mFourierSubalgebra d).topologicalClosure = ⊤ :=
+  ContinuousMap.starSubalgebra_topologicalClosure_eq_top_of_separatesPoints _
+    mFourierSubalgebra_separatesPoints
+
+/-- The linear span of the monomials `fourier n` is dense in `C(UnitAddTorus d, ℂ)`. -/
+theorem span_mFourier_closure_eq_top [DecidableEq d] :
+    (span ℂ (range <| mFourier (d := d))).topologicalClosure = ⊤ := by
+  rw [← mFourierSubalgebra_coe]
+  exact congr_arg (Subalgebra.toSubmodule <| StarSubalgebra.toSubalgebra ·)
+    mFourierSubalgebra_closure_eq_top
+
+end Algebra
+
+section Lp
+
+/-- The family of monomials `fourier n`, parametrized by `n : ℤ` and considered as
+elements of the `Lp` space of functions `AddCircle T → ℂ`. -/
+abbrev mFourierLp (p : ℝ≥0∞) [Fact (1 ≤ p)] (n : d → ℤ) :
+    Lp ℂ p (volume : Measure (UnitAddTorus d)) :=
+  ContinuousMap.toLp (E := ℂ) p volume ℂ (mFourier n)
+
+theorem coeFn_mFourierLp (p : ℝ≥0∞) [Fact (1 ≤ p)] (n : d → ℤ) :
+    mFourierLp p n =ᵐ[volume] mFourier n :=
+  ContinuousMap.coeFn_toLp volume (mFourier n)
+
+/-- For each `1 ≤ p < ∞`, the linear span of the monomials `mFourier n` is dense in the `Lᵖ` space
+of functions on `UnitAddTorus d`. -/
+theorem span_mFourierLp_closure_eq_top [DecidableEq d] {p : ℝ≥0∞} [Fact (1 ≤ p)] (hp : p ≠ ∞) :
```

Comment:
```suggestion
theorem span_mFourierLp_closure_eq_top {p : ℝ≥0∞} [Fact (1 ≤ p)] (hp : p ≠ ∞) :
```
You can remove this if you start your proof with `classical`. I guess this means many `Decidable` assumptions later on can be removed.

Reply:
Thanks! There was one `[DecidableEq d]` which I couldn't get rid of, since it's needed for the _statement_ (not just the proof) of the relevant lemma, but all the rest are now gone.